### PR TITLE
dai-zephyr: refactor dai params

### PR DESCRIPTION
**Pull request by RanderWang**
Refactor dai params into more feasible functions and remove duplicate code. No functional change.

The original pr: https://github.com/thesofproject/sof/pull/7345 was done by Jaska but was not merged. Since he has leaved intel, I will help to rebase it and submit it again .

### CODE REVIEW

**Review by btian1 — COMMENTED** [review]
the code already mismatched, if possible, I would suggest to split with several patches for better check and bisect.

**Comment by RanderWang** [comment]
All tests pass!  the original PR got 3 approve. Thanks!

File: src/audio/dai-zephyr.c [hunk 1]
```diff
@@ -956,17 +820,64 @@ int dai_common_params(struct dai_data *dd, struct comp_dev *dev,
...
+		comp_err(dev, "dai_zephyr_params(): set dma config failed.");
+out:
+	/*
+	 * Make sure to free all allocated items, all functions
+	 * can be called with null pointer.
+	 */
+	if (err < 0) {
+		buffer_free(dd->dma_buffer);
+		dma_sg_free(&config->elem_array);
+		rfree(dd->z_config);
```
Comment (lyakh) [inline src/audio/dai-zephyr.c:877]: if I understand it correctly, `dd` is persistent, isn't it? Here in error case we free several buffers, but we keep pointers to them. Should we reset them all to `NULL`? Except `config->elem_array.elems` - it's already reset inside `dma_sg_free()`

Comment (RanderWang) [reply src/audio/dai-zephyr.c:877]: dd is allocated by dai_new and released in dai_free. If dai_common_params is failed,  dai_reset will be called and these pointer will be set to null
````c
void dai_common_reset(struct dai_data *dd, struct comp_dev *dev)
{
	struct dma_sg_config *config = &dd->config;

	dma_sg_free(&config->elem_array);
	if (dd->z_config) {
		rfree(dd->z_config->head_block);
		rfree(dd->z_config);
		dd->z_config = NULL;
	}

	if (dd->dma_buffer) {
		buffer_free(dd->dma_buffer);
		dd->dma_buffer = NULL;
	}
}
````

**Review by kv2019i — APPROVED** [review]

**Comment by btian1** [comment]
> > ggest to split with several patches for better check and bisect.
> 
> NO, it is based on latest code

mismatch means code are mismatched, hard to review and revert, after this patch merged, and then following dai code changed continue, once you want to revert this patch due to a regression bug based on latest sof code, it will be hard, have to use reset.

if possible, split will make easy to read/review/maintain.

File: src/audio/dai-zephyr.c (hunk 1 above)
Comment (lyakh) [reply src/audio/dai-zephyr.c:877]: @RanderWang that's even worse then. That would mean that we free the same pointers twice, which is wrong and can lead to run-time errors

**Review by lyakh — CHANGES_REQUESTED** [review]
double free is a bug

File: src/audio/dai-zephyr.c (hunk 1 above)
Comment (RanderWang) [reply src/audio/dai-zephyr.c:877]: o, thanks!

**Comment by RanderWang** [comment]
fixed the double free, thanks!

**Review by lyakh — APPROVED** [review]

**Comment by kv2019i** [comment]
The smart-amp stub build fail is unrelated, proceeding with merge.

**Pull request merged**